File: agents/trace_pattern.py
"""痕跡データパターン分析AI（62）— 蓄積された痕跡(traces)の傾向をルールベースで集計する。
自治体向けレポート商品の中身（数字）を作る装置。LLM APIは使わずSupabase REST読み取りのみ。
投稿時間帯・また来たい率・話したい率・実験回ごとの数・書き込みの厚み（3つの問いの記入率）を出す。
"""
import json
import urllib.request
from collections import Counter
from datetime import datetime

import common


def _get(url: str, key: str, path: str) -> list:
    req = urllib.request.Request(
        f"{url}/rest/v1/{path}",
        headers={"apikey": key, "Authorization": f"Bearer {key}"},
    )
    with urllib.request.urlopen(req, timeout=15) as res:
        return json.loads(res.read())


def main():
    with common.running("trace_pattern"):
        env = common.load_env_local()
        url = env.get("NEXT_PUBLIC_SUPABASE_URL")
        key = env.get("SUPABASE_SERVICE_ROLE_KEY") or env.get("NEXT_PUBLIC_SUPABASE_ANON_KEY")
        if not url or not key:
            common.write_result("trace_pattern", {"error": "Supabase設定(.env.local)が見つかりません"})
            return
        try:
            rows = _get(url, key,
                        "traces?select=created_at,why,interpretation,self_reflection,want_revisit,want_to_share,session_code,nickname&limit=5000")
        except Exception as e:
            common.write_result("trace_pattern", {"error": f"取得エラー: {e}"})
            return

        total = len(rows)
        if total == 0:
            common.write_result("trace_pattern", {"total": 0, "note": "まだ痕跡がありません"})
            return

        hours = Counter()
        revisit = share = 0
        deep = 0  # 3つの問いを全部書いた濃い痕跡
        sessions = Counter()
        for r in rows:
            ca = r.get("created_at", "")
            try:
                hours[datetime.fromisoformat(ca.replace("Z", "+00:00")).hour] += 1
            except Exception:
                pass
            if r.get("want_revisit"):
                revisit += 1
            if r.get("want_to_share"):
                share += 1
            if (r.get("why") and r.get("interpretation") and r.get("self_reflection")):
                deep += 1
            if r.get("session_code"):
                sessions[r["session_code"]] += 1

        peak_hours = [f"{h}時({n})" for h, n in hours.most_common(3)]
        common.write_result("trace_pattern", {
            "total": total,
            "want_revisit_rate": round(revisit / total * 100, 1),
            "want_to_share_rate": round(share / total * 100, 1),
            "deep_write_rate": round(deep / total * 100, 1),
            "peak_hours": peak_hours,
            "top_sessions": sessions.most_common(5),
        })


if __name__ == "__main__":
    main()

File: lib/tracePattern.ts
// ============================================================
// 痕跡データパターン分析：agents/trace_pattern.py（番人62）と同じ集計を
// サイト本体からライブに読む。自治体向けレポート商品の中身（数字）を作る装置。
// 投稿時間帯のピーク・また来たい率・話したい率・書き込みの厚み（3問完答率）・
// 実験回別の件数を返す。個人を特定できる値は含めない。
// ============================================================
import type { SupabaseClient } from '@supabase/supabase-js';

interface TraceRow {
  created_at: string;
  why: string | null;
  interpretation: string | null;
  self_reflection: string | null;
  want_revisit: boolean;
  want_to_share: boolean;
  session_code: string | null;
}

export interface TracePatternResult {
  ok: boolean;
  generatedAt: string;
  total: number;
  wantRevisitRate?: number;
  wantToShareRate?: number;
  deepWriteRate?: number;
  peakHours?: { hour: number; count: number }[];
  topSessions?: { sessionCode: string; count: number }[];
  error?: string;
  note?: string;
}

export async function computeTracePattern(supabaseServer: SupabaseClient): Promise<TracePatternResult> {
  const generatedAt = new Date().toISOString();
  const { data, error } = await supabaseServer
    .from('traces')
    .select('created_at, why, interpretation, self_reflection, want_revisit, want_to_share, session_code')
    .eq('is_deleted', false)
    .limit(5000);

  if (error) return { ok: false, generatedAt, total: 0, error: error.message };

  const rows = (data ?? []) as TraceRow[];
  const total = rows.length;
  if (total === 0) return { ok: true, generatedAt, total: 0, note: 'まだ痕跡がありません' };

  const hourCounts = new Map<number, number>();
  const sessionCounts = new Map<string, number>();
  let revisit = 0, share = 0, deep = 0;

  for (const r of rows) {
    const hour = new Date(r.created_at).getUTCHours();
    if (Number.isFinite(hour)) hourCounts.set(hour, (hourCounts.get(hour) ?? 0) + 1);
    if (r.want_revisit) revisit++;
    if (r.want_to_share) share++;
    if (r.why && r.interpretation && r.self_reflection) deep++;
    if (r.session_code) sessionCounts.set(r.session_code, (sessionCounts.get(r.session_code) ?? 0) + 1);
  }

  const peakHours = [...hourCounts.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, 3)
    .map(([hour, count]) => ({ hour, count }));

  const topSessions = [...sessionCounts.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5)
    .map(([sessionCode, count]) => ({ sessionCode, count }));

  return {
    ok: true,
    generatedAt,
    total,
    wantRevisitRate: Math.round((revisit / total) * 1000) / 10,
    wantToShareRate: Math.round((share / total) * 1000) / 10,
    deepWriteRate: Math.round((deep / total) * 1000) / 10,
    peakHours,
    topSessions,
  };
}
